### PR TITLE
Ssao

### PR DESCRIPTION
**Pull request by mchaptel**
This is a big one, implementing SSAO for NVR

some screenshots:

before:

after:

some more:

It's a modified version of the Crytech SSAO with additions that take inspiration from the Starcraft 2 methods.

### CODE REVIEW

**Comment by mchaptel** [comment]
Would it be ok to not squash commits? I made sure to separate them in ways that make them easily revertable

**Comment by llde** [comment]
Seems fine, a comment about the difference between ProjMatrix and RealPRojMatrix would be nice.

Also why do you calcuklate the Depth Reconstruction Constant in the Cpu? Woudln't be better to  calculate them in the shader?

**Comment by mchaptel** [comment]
> Seems fine, a comment about the difference between ProjMatrix and RealPRojMatrix would be nice.

I've based mine on [this](https://learn.microsoft.com/en-us/windows/win32/direct3d9/projection-transform), as per this comment:
https://github.com/llde/TESReloaded10/pull/74/files#diff-42a641c90c3d3aae861e93c5a5daaa6d7ad67fbd06983049c025a2506d8674a9R92

Not completely sure how the old one was built. I've left it because I don't quite understand it and the one I used might very well be wrong... at least it seems to let me project and get comparable values from the depth buffer

> Also why do you calcuklate the Depth Reconstruction Constant in the Cpu? Woudln't be better to calculate them in the shader?
These constants are the same for all pixels, and I think they're also used by a bunch of different shaders, so to me it made sense to only compute them once. I also think eventually these would be nice to add to an include, so all shaders use the same system, what do you think?

**Comment by llde** [comment]
> These constants are the same for all pixels, and I think they're also used by a bunch of different shaders, so to me it made sense to only compute them once. I also think eventually these would be nice to add to an include, so all shaders use the same system, what do you think?
In the vertex part they are calculated once for vertex (4 times in Effects), and it's nelgigible respect to the gain on calcualting on the gpu and avoid a RAM/GPU transfer
But well I don't expect  differences at all, considering that we already pass memory around



> Not completely sure how the old one was built. I've left it because I don't quite understand it and the one I used might very well be wrong... at least it seems to let me project and get comparable values from the depth buffer

Does uing the new one provide a difference? I believe so considering you spent time on it.

**Comment by mchaptel** [comment]
> In the vertex part they are calculated once for vertex (4 times in Effects), and it's nelgigible respect to the gain on calcualting on the gpu and avoid a RAM/GPU transfer But well I don't expect differences at all, considering that we already pass memory around Agree for the include

That's interesting! I have to say that the DepthConstants vector is mostly old constants used by the previous depth reconstruction system. All the new code uses the CameraData and does these calculations on the shader. Anyway I don't know much about these things so thanks for the info!

> Does using the new one provide a difference? I believe so considering you spent time on it.

Basically the SSAO algorithm requires to:
1 - recompose view position from screen coordinates + depth
2 - add a vector to this position in view space
3 - reproject this position into screenspace and sample the depthbuffer at the same coordinates, then compare values.

At the beggining, I was having a lot of trouble with getting comparable values, both in depth reconstruction and projecting back my point and getting the same results. I had to reverse engineer a way to use the projection matrix and understand the way the depth buffer is encoded to get acceptable values, and those matrices are the result of that. There is code in the previous version that does similar things, but I couldn't figure out how to get useable data from it. I was also surprised to notice that the depth reconstruction function from the Shadows shader and from the AO shader were different, so one of them is probably wrong.. 

Anyway they must have been created this way for a reason in the first place, I just don't know enough to know what that reason is. I thought using methods found in most available tutorials would be easier to understand for other people in the future.

**Comment by llde** [comment]
> I was also surprised to notice that the depth reconstruction function from the Shadows shader and from the AO shader were different, so one of them is probably wrong..

You should try to see what happens if you change the position recostruction from depth (respecting the different coordinates systems)  of the shadows with the new matrix.

**Comment by mchaptel** [comment]
> You should try to see what happens if you change the position recostruction from depth (respecting the different coordinates systems) of the shadows with the new matrix.

That was going to be my next steps. Basically, the shadows do something weird:
  * projects the point from screen position into the world coordinates
  * projects back the point to camera space but now it has a depth
  * then it uses a premultiplied matrix that transforms from screen space to light space.
  
If the matrix was working, we should just be able to transform to world, then from world to light space or something. That's why I think something might be off as well. 

Anyway that's going to be a lot of work and I think outside the scope of this PR. But I am planning on doing it.

**Comment by mchaptel** [comment]
I've plugged the new matrix into the shadow calculations and they work fine. I need to check by doing the comparison in world space without transforming back into screenspace to fully check.

It seems assuming the FOV is actually the vertical FOV also fixes the cascade corners issue, even without using a safety value.
I would like to add it a setting for display distance for the AO before merging.

**Comment by llde** [comment]
The FOV must be checked for Oblivion too, and in case there is a difference add a mechanism to automatically select the correct one (possibily without needing a branch in the effect)

Regarding the matrix instead is supposed to be "plug and play" or it require a different code to restructure the position?

**Comment by mchaptel** [comment]
Yeah checking for Oblivion would be good, although I don't actually own Oblivion to check. 

That said, this is a standard projection matrix, not anything fallout related, so it shouldn't be a problem.

It's plug & play right now, of course you still need to divide by w and shift coordinates from 0 - 1 to -1 to 1 as usual, but apart from that it seems to work according to what most ressources on the subject describe. You can see the functions in the AO shader to reconstruct position.

**Comment by llde** [comment]
What I meant for the last point is: If I change globally the matrix to the new one without chinging other code will it work or it expect something different?

**Comment by mchaptel** [comment]
actually I think it should be ok, but it requires maybe more thorough testing. I've swapped it at the moment and it seems to work fine.

**Comment by llde** [comment]
Ok.

**Comment by mchaptel** [comment]
> Ok.

Is there anything you want me to fix/change before this PR can be merged? Let me know. I'm happy to wait for it to be merged before changing the settings.

**Comment by mchaptel** [comment]
I've reverted the extra matrix, after further testing the main issue was the depth reconstruction functions and not the matrix itself!

**Comment by llde** [comment]
Sorry I wasn't that much present on discord these days, Was the new matrix having some issues?

File: TESReloaded/Core/RenderManager.cpp
Comment (llde) [inline TESReloaded/Core/RenderManager.cpp:185]: What's  changing projMatrix._33 from Frustum->Far * InvFmN to farZ / (farZ - nearZ) do?
Also the cascaded changes to projMatrix._43

**Review by llde — CHANGES_REQUESTED** [review]
Remove Game.h changes, tracked in another PR.
Copy the new effect to Oblivion Extra files.

File: TESReloaded/Core/RenderManager.cpp
Comment (mchaptel) [reply TESReloaded/Core/RenderManager.cpp:185]: it's the same thing, just less obscured and closer in formulation to the type of syntax that can be found in ressources on the subject on the net

File: TESReloaded/Core/RenderManager.cpp
Comment (mchaptel) [reply TESReloaded/Core/RenderManager.cpp]: when figuring out the position reconstruction, I was having issues so I reimplemented a projection matrix based on ressources found on the subject to make sure I was doing everything by the book.

After I was having issues in screen space shadows, I discovered that the original matrix, although weird, was in fact correct so I've reverted the changes and removed my experimental matrix. The issues lied in the position reconstruction functions instead

**Comment by mchaptel** [comment]
About the changes - can do, I'll cherry pick the relevant commits and resubmit. This branch is rebased on the `fix underwater` branch which I assumed would be easier to merge first

File: TESReloaded/Core/RenderManager.cpp
Comment (llde) [reply TESReloaded/Core/RenderManager.cpp]: Wait, was the issue in the screen space shadows related to position recostruction or the new matrix? Did you fiugre out how the matrix differed?

Comment (mchaptel) [reply TESReloaded/Core/RenderManager.cpp]: The new matrix was removed and the old one restored. The issue was in the obscure way there components were described as well as the reconstruction function itself.

**Comment by mchaptel** [comment]
I've pushed a new branch with same changes but cleaner history

https://github.com/llde/TESReloaded10/pull/79

**Comment by llde** [comment]
Merged #79

**Pull request closed without merging**
